src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,553 @@
+  """
+  
+  # Get all the grants inside neighborhood of the protection entity
+  grants_inside = interf.findGrantsInsideNeighborhood(
+                    grant_objects, protection_point, protection_ent_type)
+  
+  if len(grants_inside) > 0:
+
+    for channel in channels:
+      # Set interference quota for protection_point, channel combination
+
+      # Calculate Roll of Attenuation for ESC Sensor Protection entity
+      if protection_ent_type is interf.ProtectedEntityType.ESC:
+        if channel[0] >= 3650.e6:
+          center_freq = (channel[0] + channel[1]) / 2
+          interference_quota = interf.dbToLinear(interf.linearToDb(threshold) - 
+                (2.5 + ((center_freq - interf.ESC_CH21_CF_HZ) / interf.MHZ)))
+        else:
+          interference_quota = threshold
+
+      # Get the FSS blocking passband range
+      elif protection_ent_type is interf.ProtectedEntityType.FSS_BLOCKING:
+        interference_quota = threshold
+        channel = [low_freq, high_freq]
+
+      else:
+        interference_quota = threshold
+
+      # Get protection constraint over 5MHz channel range
+      channel_constraint = interf.ProtectionConstraint(latitude=protection_point[0],
+                             longitude=protection_point[1], low_frequency=channel[0],
+                             high_frequency=channel[1], entity_type=protection_ent_type)
+
+      # Identify CBSD grants overlapping with frequency range of the protection entity 
+      # in the neighborhood of the protection point and channel
+      neighborhood_grants = interf.findOverlappingGrants(
+                              grants_inside, channel_constraint)
+      num_grants = len(neighborhood_grants)
+
+      if not neighborhood_grants:
+        continue
+
+      # calculate the fair share
+      fair_share = float(interference_quota) / num_grants
+
+      # Empty list of final IAP EIRP and interference assigned to the grant
+      grants_eirp = []
+
+      # Initialize interference list to 0 for all the neighborhood_grants
+      interference_list = [0] * num_grants
+
+      aggr_interference = 0
+
+      # Initialize flag to False for all the grants
+      grant_initialize_flag = [False] * num_grants
+
+      # Initialize grant EIRP to maxEirp for all the grants 
+      grants_eirp = [grant.max_eirp for grant in neighborhood_grants]
+
+      while (num_grants > 0):
+        for index, grant in enumerate(neighborhood_grants):
+          if grant_initialize_flag[index] is False:
+            # Compute interference grant causes to protection point over channel
+            interference = interf.computeInterference(grant, grants_eirp[index], 
+                             channel_constraint, fss_info, esc_antenna_info, region_type)
+            # calculated interference exceeds fair share of
+            # interference to which the grants are entitled
+            if interference < fair_share:
```

Comment:
Current   logic is as per your comment.           While loop for a particular 5MHz channel is executed for all the CBSDs,   while loop executes till all the CBSd power is reduced to meet margin   criteria.           Same logic is repeated for next 5MHz channel

Reply:
For a given CBSD, if its grant is spanning multiple 5 MHz channels, then it should not be considered satisfied until the interference from all of the 5MHz channels that it is overlapping satisfies the 1/N_ch criteria. Only then will the CBSDs interference be removed from the respective 5 MHz channels. Meaning the IAP is done jointly over the 5MHz channels that the grant is spanning. Here the CBSD margins are computed independently over each of the 5MHz channels and not jointly estimated over the frequency channels

Reply:
We are investigating this point internally, about the joint vs independent spanning of 5MHz channel.
Will give our take on this soon here

Reply:
Navin, 
after double checking internally, we do not agree with your point here. Conceptually this could have been one approach (more complicated), but the "SAS Provider WG Agreement" says otherwise:
>> The joint IAP Protection is provided over each 5MHz segment of the whole frequency channel 
>> occupied (or licensed) by the incumbent or PAL, assuming that each 5 MHz segment is a 
>> separate constraint/channel point. 

Basically the whole point of the IAP process is *not* to effectively assign actual power to the CBSD, but instead to find the sharing of the the interference allowance between the SAS for each point and channel.
 
Then the SAS can distribute power among their managed CBSD within this allowance, possibly using a process similar to the IAP but taking into account the most constraining channel in consideration (or some other process). 
But this is out of scope of the IAP reference model here.

Reply:
We have discussed this in our SAS admin meeting quite extensively  and as captured in the agreement the 5 MHz segments are  considered jointly for a protection point. I agree that each constraint consists of a 5 MHz segment (overlapping the incumbent)  and  a protection point.